replica/tablets.cc

```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
+    return for_each_sstable_set_until(query::full_partition_range, [func = std::move(func)] (lw_shared_ptr<sstables::sstable_set> set) {
+        return set->for_each_sstable_until(func);
+    });
+}
+
+future<stop_iteration> tablet_sstable_set::for_each_sstable_gently_until(std::function<future<stop_iteration>(const sstables::shared_sstable&)> func) const {
+    return for_each_sstable_set_gently_until(query::full_partition_range, [func = std::move(func)] (lw_shared_ptr<sstables::sstable_set> set) {
+        return set->for_each_sstable_gently_until(func);
+    });
 }
+
+bool tablet_sstable_set::insert(sstables::shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+bool tablet_sstable_set::erase(sstables::shared_sstable sst) {
+    throw_with_backtrace<std::bad_function_call>();
+}
+
+class tablet_incremental_selector : public sstables::incremental_selector_impl {
+    const tablet_sstable_set& _tset;
+
+    // _cur_set and _cur_selector contain a snapshot
+    // for the currently selected compaction_group.
+    lw_shared_ptr<sstables::sstable_set> _cur_set;
+    std::optional<sstables::sstable_set::incremental_selector> _cur_selector;
+    dht::token _lowest_next_token = dht::maximum_token();
+
+public:
+    tablet_incremental_selector(const tablet_sstable_set& tset)
+            : _tset(tset)
+    {}
+
+    virtual std::tuple<dht::partition_range, std::vector<sstables::shared_sstable>, dht::ring_position_ext> select(const dht::ring_position_view& pos) override {
+        // Always return minimum singular range, such that incremental_selector::select() will always call this function,
+        // which in turn will find the next sstable set to select sstables from.
+        const dht::partition_range current_range = dht::partition_range::make_singular(dht::ring_position::min());
+
+        if (!_cur_selector) {
+            auto idx = _tset.group_of(pos.token());
+            _cur_set = _tset._sstable_sets[idx];
+            // Set the next token to point to the possible storage group.
+            // It will be considered later on when the _cur_set is exhausted
+            _lowest_next_token = find_lowest_next_token(_tset.first_token_of(idx+1));
+            if (!_cur_set) {
+                auto lowest_next_position = _lowest_next_token.is_maximum()
+                    ? dht::ring_position_ext::max()
+                    : dht::ring_position_ext::starting_at(_lowest_next_token);
+                tablet_logger.debug("tablet_incremental_selector {}.{}: select pos={}: returning 0 sstables, next_pos={}",
+                        _tset.schema()->ks_name(), _tset.schema()->cf_name(), pos, lowest_next_position);
+                return std::make_tuple(std::move(current_range), std::vector<sstables::shared_sstable>{}, lowest_next_position);
+            }
+
+            _cur_selector.emplace(_cur_set->make_incremental_selector());
+        }
+
```

Comment:
we need a maybe_advance_selector() here, before we select below.

otherwise, we'll query a token that belongs to next group on the current one.

select() is called again when a next token is found that >= lowest_next_token, so we cannot advance selector after select() call below.

I think it's important that we add tests for such edge cases, when range scan is jumping across tablet boundaries for example. we want to make sure that data is properly returned.

e.g. 

read last token from tablet 0
read first token from tablet 1.